ext/git.c

```diff
@@ -0,0 +1,388 @@
+        if (!zend_string_equals(git_metadata->property_commit, commit_sha)) {
+            zend_string *repository_url = get_repository_url(ZSTR_VAL(git_dir));
+            replace_git_metadata(git_metadata, commit_sha, repository_url);
+        } else {
+            zend_string_release(commit_sha);
+        }
+    } else if (commit_sha) {
+        zend_string *repository_url = get_repository_url(ZSTR_VAL(git_dir));
+        replace_git_metadata(git_metadata, commit_sha, repository_url);
+    }
+
+    zend_string_release(git_dir);
+}
+
+void update_git_metadata(void) {
+    zend_string *cwd;
+    git_metadata_t *git_metadata;
+    ZEND_HASH_FOREACH_STR_KEY_PTR(&DDTRACE_G(git_metadata), cwd, git_metadata) {
+        refresh_git_metadata_if_needed(cwd, git_metadata);
+    }
+    ZEND_HASH_FOREACH_END();
+}
+
+bool inject_from_env() {
+    zend_string *commit_sha = get_DD_GIT_COMMIT_SHA();
+    zend_string *repository_url = get_DD_GIT_REPOSITORY_URL();
+    return add_git_info(commit_sha, repository_url);
+}
+
+bool inject_from_global_tags() {
+    bool success = false;
+    zend_array *global_tags = get_DD_TAGS();
+    zval *commit_sha = zend_hash_str_find(global_tags, ZEND_STRL("git.commit.sha"));
+    zval *repository_url = zend_hash_str_find(global_tags, ZEND_STRL("git.repository_url"));
+    if (commit_sha && repository_url && Z_TYPE_P(commit_sha) == IS_STRING && Z_TYPE_P(repository_url) == IS_STRING) {
+        success = add_git_info(Z_STR_P(commit_sha), Z_STR_P(repository_url));
+    }
+    return success;
+}
+
+bool inject_from_git_dir() {
+    zend_string *cwd = get_current_working_directory();
+    if (!cwd) return false;
+
+    update_git_metadata();
+
+    git_metadata_t *git_metadata = zend_hash_find_ptr(&DDTRACE_G(git_metadata), cwd);
+    if (git_metadata) {
+        use_cached_metadata(git_metadata);
+        zend_string_release(cwd);
+        return true;
+    }
+
+    zend_string *git_dir = find_git_dir(ZSTR_VAL(cwd));
+    if (!git_dir) {
+        zend_string_release(cwd);
+        return false;
+    }
+
+    bool success = process_git_info(git_dir, cwd);
+
+    zend_string_release(git_dir);
+    zend_string_release(cwd);
+
+    return success;
+}
+
+void ddtrace_inject_git_metadata(zval *carrier) {
+    if (DDTRACE_G(git_object) == &empty_git_object.std) {
+        return;
+    }
+
+    if (DDTRACE_G(git_object) || inject_from_env() || inject_from_global_tags() || inject_from_git_dir()) {
+        ZVAL_OBJ_COPY(carrier, DDTRACE_G(git_object));
+    } else {
+        DDTRACE_G(git_object) = &empty_git_object.std;
+    }
+}
+
+void ddtrace_clean_git_metadata(void) {
+    git_metadata_t *val;
+    ZEND_HASH_FOREACH_PTR(&DDTRACE_G(git_metadata), val) {
+        if (val->property_commit) zend_string_release(val->property_commit);
+        if (val->property_repository) zend_string_release(val->property_repository);
+        pefree(val, 1);
+    }
+    ZEND_HASH_FOREACH_END();
+    zend_hash_destroy(&DDTRACE_G(git_metadata));
+}
```

Comment:
Is there a reason to not passing a destructor function when initializing the `DDTRACE_G(git_metadata)` with `zend_hash_init` instead of this function?

Reply:
There was no particular reason besides that I wasn't aware of it. Added one :+1: